python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+The point of these indirections is that only string references rather than objects are passed between Matlab and Python
+but the behaviour is similar to creating objects and accessing their properties and methods directly.
+
+
+# Limitations and future work
+
+There are currently some limitations due to the architecture chosen for `pyHorace` which can be alleviated with more work:
+
+1. The Python syntax for any Matlab function/methods must exactly follow the Matlab syntax.
+2. Matlab functions can only be called as methods of the `Matlab` class and not independently (e.g. `without the m.* prefix`)
+3. Matlab help text cannot be accessed from Python.
+4. MPI routines don't work.
+
+Limitations 1 and 2 come from the choice, outlined in the [first section](#compiled-matlab) to use a gateway function which calls `feval`.
+This can be overcome by writing additional Python wrapper for each Matlab function or class we want to expose to Python.
+Alternatively the `Matlab` and `MatlabProxyObject` classes can be modified to automatically
+convert Python keyword arguments to `('name', value)` pairs to be passed to Matlab if the Matlab code is modified to accept them.
+
+Limitation 3 comes from a restriction of the compiled Matlab output itself:
+Matlab does not allow access to its online documentation system from compiled code.
+The only way to work around this is to save the help text for every Matlab function or class we want to expose to Python to a text file
+and then to either generate wrappers for each (such as that used to overcome limitations 1 and 2)
+or to modify the `Matlab` and `MatlabProxyObject` classes to look up these help text.
+
+Limitation 4 is because the MPI routines needs to launch additional instances of the Matlab interpreter which is not possible from within Python.
+The most straight-forward way to overcome this limitation is to compile Horace as a stand-alone executable, with the `worker_v2.m` function as the gateway function. 
+This executable can then be launched and will communicate with the main process using files or MPI as usual. (This has not been tested).
+
+In addition, however, there are also limitations imposed by Matlab which cannot be alleviated:
+
+5. Must compile Matlab code for specific Matlab release.
+6. Must use a specific C++ compiler for the `mex` files.
+
+Restriction 5 means for example that a package created with Matlab R2020a can only be used with the MCR 2020a version.
+(Note that users must download and install the MCR to run `pyHorace` even if they alread have a Matlab license and Matlab installed
+unless they also have the Matlab Compiler toolbox installed).
+Matlab also only support specific C++ compilers for `mex` files.
+Using an older compiler will generate files which will work on a new Matlab version but the reverse is generally not true.
+This means that if we want to support the widest range of operating systems and Matlab versions,
```

Comment:
The Horace build pipelines are creating and testing mex files with multiple MATLAB/platform combination and making all resulting binaries available on the GitHub releases page.

Reply:
Yeah... the only issue is that Matlab is a few releases behind the latest compilers (especially on Linux), so e.g. won't be able to use some C++17 or C++20 features until a few years time.